src/common/utils/pipeline.ts

```diff
@@ -1,17 +1,62 @@
-import { AxiosInstance } from 'axios';
+import AppService from '../../app-config/service';
+import { StringTemplateBuilder, TemplateValues } from './template-builder';
+
+class PipelineAbortedError extends Error {
+  constructor(deployment_id: string, deployment_link: string) {
+    super();
+
+    this.message = `Deployment ${deployment_id} was aborted\n See deployment log for more details: ${deployment_link}`;
+  }
+}
+
+class DeploymentFailedError extends Error {
+  constructor(pipeline_id: string, deployment_links: string[]) {
+    super();
+
+    const deployment_string = deployment_links.length > 1
+      ? `${deployment_links.length} deployments`
+      : '1 deployment';
+    const listified_link_string = deployment_links.map((s: string) => `- ${s}`).join('\n');
+    this.message = `Pipeline ${pipeline_id} failed because ${deployment_string} failed:\n${listified_link_string}`;
+  }
+}
+
+const deploymentUrlBuilder = new StringTemplateBuilder(
```

Comment:
I'm not convinced we need a string template builder. Is the problem that we don't want to hardcode api endpoints  in the cli?

Feels like more code to define the same url:
```ts
const environment_url = `${app_host}/${account.name}/environments/${environment.name}`;
const deployment_url = `${environment_url}/deployments/${deployment.id}`;
```

If we wanted to take out the guess work we could create controller methods like the cloud/api:
`app.api.pipelines.getDeployments()`
or helpers to get a url
`app.cloud.deployments.getUrl(deployment)`

Reply:
The problem was reducing code repetition and improving readability. In the case of the deployment urls, there are 4 variables that need to be templated, which got relatively messy to write out inline:

```
const deploymentUrl = `${app.config.app_host}/${pipeline.environment.account.name}/environments/${pipeline.environment.name}/deployments/${aborted_deployments[0].id}`
```

Abstracting this out to a function helped a bit:

```
const buildDeploymentUrl = (app_host, account, environment, deployment) =>
  `${app_host}/${account}/environments/${environment}/deployments/${deployment}`;

...

const deploymentUrl = buildDeploymentUrl(
  app.config.app_host,
  pipeline.environment.account.name,
  pipeline.environment.name
  aborted_deployments[0].id
);
```

But this is still a lengthy function call, and we need to build this URL in multiple spots which added a lot of visual heft to the `pollPipeline` function overall.

Between the two calls though the only thing that was changing was the final deployment id, which made me want to just curry those first 3 function inputs. Though the template builder is using the builder pattern, it achieves the same ergonomics as currying in this case, letting us set the host, account and env name just once and then build various deployment urls as needed without having to repeat inputs

Reply:
Yeah I definitely agree that its ugly in a one liner. Still not sold on the need for string templating since typescript does that already via `${...}`. 

I'd rather hide away the building of the urls.  That way a dev doesn't need to know how to build. Something like:
```ts
app.cloud.deployments.getUrl(deployment);

getUrl(deployment: Deployment) {
  const environment = deployment.pipeline.environment;
  if (environment) {
    return `${this.config.app_host}/${environment.account.name}/environments/${environment.name}/deployments/${deployment.id}`
  } else if (platform) {
    return ...
  }
}
```